src/Adapter/ContextStateManager.php

```diff
@@ -0,0 +1,192 @@
+<?php
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter;
+
+use Cart;
+use Context;
+use Country;
+use Currency;
+use Customer;
+use Language;
+
+/**
+ * Allows manipulating context state.
+ * This was adapted for specific broken legacy use cases when the previous state of context must be restored after some actions.
+ *
+ * e.g. order creation from back office.
+ *  Legacy requires Context properties (currency, country etc.) instead of using cart properties
+ *  so some context props must be changed for a while and then restored to previous state.
+ */
+final class ContextStateManager
+{
+    /**
+     * @var Context
+     */
+    private $context;
+
+    /**
+     * @var Cart|null
+     */
+    private $cart;
+
+    /**
+     * @var Country|null
+     */
+    private $country;
+
+    /**
+     * @var Currency|null
+     */
+    private $currency;
+
+    /**
+     * @var Language|null
+     */
+    private $language;
+
+    /**
+     * @var Customer|null
+     */
+    private $customer;
+
+    /**
+     * @var bool
+     */
+    private $cartStateChanged = false;
```

Comment:
The use of boolean is redundant, you simply check that the class fields are not null any more

Reply:
in context it is possible that cart will be null, so we cannot rely on null check, that's why i made those explicit checks.

Reply:
ooh nice catch!